reflex/utils/types.py

```diff
@@ -320,24 +358,9 @@ def get_attribute_access_type(cls: GenericType, name: str) -> GenericType | None
     if hint := get_property_hint(attr):
         return hint
 
-    if (
-        hasattr(cls, "__fields__")
-        and name in cls.__fields__
-        and hasattr(cls.__fields__[name], "outer_type_")
-    ):
+    if hasattr(cls, "__fields__") and name in cls.__fields__:
         # pydantic models
-        field = cls.__fields__[name]
-        type_ = field.outer_type_
-        if isinstance(type_, ModelField):
-            type_ = type_.type_
-        if (
-            not field.required
-            and field.default is None
-            and field.default_factory is None
-        ):
-            # Ensure frontend uses null coalescing when accessing.
-            type_ = Optional[type_]
-        return type_
```

Comment:
i'm really hesitant about removing these lines, unless we're sure this is not going to be hit.

@benedikt-bartscher do you know under what circumstances this bit of code gets used?

Reply:
```python
import reflex as rx


class M1(rx.Base):
    foo: str = "foo"


class M2(rx.Base):
    m1: M1 = None


class State(rx.State):
    m2: M2 = M2()


def index() -> rx.Component:
    return rx.container(
        rx.text(State.m2.m1.foo),
    )


app = rx.App()
app.add_page(index)
```

This is at least one of the cases; default is `None`, but the field is not annotated as `Optional` (which is incorrect and i'm surprised pydantic doesn't yell), but Reflex still try to be helpful and use "null coalescing" to avoid frontend exceptions.

Again, why pydantic allows the default to not match the annotation and pass validation i'm unsure.

Reply:
@masenf sorry, I did not see your question for some reason. But it seems like you found a case.

> This one is interesting, default is None when it's unset and when it's set to None :(

@adhami3310 you can use this to determine unset defaults

```python
from pydantic.v1 import BaseModel


class User(BaseModel):
    no_default: str | None
    none_default: str | None = None


no_default_field = User.__fields__["no_default"]
none_default_field = User.__fields__["none_default"]
print(no_default_field.field_info.default)
print(none_default_field.field_info.default)
```

Reply:
TIL `PydanticUndefined`

Reply:
The question is: should we migrate everything from ModelField to FieldInfo?

Reply:
last time i checked, `FieldInfo` doesn't contain the type, but I wouldn't be surprised if i missed something

Reply:
I mean for default and default factory values.

Reply:
i made the check for default use None, there's little I can do with default_factory returning None (i hope no one thinks that's a good idea)

Reply:
Are you sure this is a bad idea?

```python
diff --git a/reflex/state.py b/reflex/state.py
index e7e6bcf3..46ba3133 100644
--- a/reflex/state.py
+++ b/reflex/state.py
@@ -1099,7 +1099,10 @@ class BaseState(Base, ABC, extra=pydantic.Extra.allow):
         if (
             not field.required
             and field.default is None
-            and field.default_factory is None
+            and (
+                field.default_factory is None
+                or types.default_factory_can_be_none(field.default_factory)
+            )
             and not types.is_optional(prop._var_type)
         ):
             # Ensure frontend uses null coalescing when accessing.
diff --git a/reflex/utils/types.py b/reflex/utils/types.py
index b8bcbf2d..60eb456d 100644
--- a/reflex/utils/types.py
+++ b/reflex/utils/types.py
@@ -333,7 +333,10 @@ def get_attribute_access_type(cls: GenericType, name: str) -> GenericType | None
         if (
             not field.required
             and field.default is None
-            and field.default_factory is None
+            and (
+                field.default_factory is None
+                or default_factory_can_be_none(field.default_factory)
+            )
         ):
             # Ensure frontend uses null coalescing when accessing.
             type_ = Optional[type_]
@@ -893,3 +896,18 @@ def typehint_issubclass(possible_subclass: Any, possible_superclass: Any) -> boo
         for provided_arg, accepted_arg in zip(provided_args, accepted_args)
         if accepted_arg is not Any
     )
+
+
+def default_factory_can_be_none(default_factory: Callable) -> bool:
+    """Check if the default factory can return None.
+
+    Args:
+        default_factory: The default factory to check.
+
+    Returns:
+        Whether the default factory can return None.
+    """
+    type_hints = get_type_hints(default_factory)
+    if hint := type_hints.get("return"):
+        return is_optional(hint)
+    return default_factory() is None
```

Reply:
most default factories are lambdas so i would imagine you're not getting a type hint in most cases, calling it could give you should be a reliable indicator but i'm afraid of side effects or performance costs